### PR TITLE
[clang-offload-bundler] Add new unbundling mode 'a' for archives

### PR DESCRIPTION
**Pull request by sndmitriev**
This patch adds support for unbundling archives into archives under a new
file type 'a'. Input file for this mode is expected to be an archive with
fat object files and outputs (one per offload target) will be archives
with extracted device specific parts from the input's objects.

Signed-off-by: Sergey Dmitriev <serguei.n.dmitriev@intel.com>

### CODE REVIEW

**Review by elizabethandrews — COMMENTED** [review]
From a quick look, I don't see anything wrong. But I am also unfamiliar with the offload-bundler and so am not the right person to review this.

File: clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp
Comment (bader) [inline clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp:965]: Can we replace `FileList` with `Archive`?
I'm not suggesting doing it right away, but I don't understand the difference between these types, which can justify having both of them. Considering that `Archive` is already supporting in llorg, it seems reasonable to replace `FileList` with `Archive`. What do you think?

**Review by bader — APPROVED** [review]

File: clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp
Comment (mdtoguchi) [reply clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp:965]: I'm sure once we change from `aoo` to `a` usage in the driver, we can get rid of the FileList type.

**Review by RaviNarayanaswamy — COMMENTED** [review]
Are these changes same as what the llvm openmp is planning to check into open source

File: clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp
Comment (sndmitriev) [reply clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp:965]: Yes, that is the plan.

**Comment by sndmitriev** [comment]
> Are these changes same as what the llvm openmp is planning to check into open source

No, this is a different change. We have added support for unbundling archives in sycl branch long time ago, but we did not have ‘a’ mode where the extracted target objects are expected to be packed back into archives. So, now I am just adding this missing part.

File: clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp
Comment (sndmitriev) [reply clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp:1141]: > Are we not doing this for Windows? Maybe add a comment why?

Windows static libraries are archives with K_GNU kind.

> Also I noticed there is K_GNU64 (and K_DARWIN64) as well. I wonder if K_GNU64 should be used for SYCL targets, which are 64-bit.

Archive writer will change 32-bit kind to 64-bit if necessary, so there is no need to use 64-bit kind explicitly.

**Review by kbobrovs — APPROVED** [review]

**Comment by sndmitriev** [comment]
I believe LIT fails in checkin testing are not related to these changes. The original patch has successfully passed testing, and LIT failures appeared after correcting typo in comments.

**Comment by sndmitriev** [comment]
I think I have addressed all comments now. Is there anything else that need to be done to get this merged?

**Comment by bader** [comment]
"Waiting on code owner review from AGindinson and/or mdtoguchi."

File: clang/test/Driver/clang-offload-bundler.c
Comment (AGindinson) [inline clang/test/Driver/clang-offload-bundler.c:347]: Just to make sure I understand, why exactly does the archive end up being the same size as the extracted host object?

Comment (sndmitriev) [reply clang/test/Driver/clang-offload-bundler.c:347]: The host bundle is handled in a special way, we just copy input archive as is to the output file. Please see ClangOffloadBundle.cpp, lines 1044-1048.

Comment (AGindinson) [reply clang/test/Driver/clang-offload-bundler.c:347]: I see, thanks! What perplexes me here is that I thought the driver's archive-based flow to be sending the unbundled host objects right to the linker, which in turn would expect (?) them to be "true" C++ object files and not the "fat" ones. Is this being changed somehow, or am I misinterpreting the flow/change itself?

Comment (sndmitriev) [reply clang/test/Driver/clang-offload-bundler.c:347]: No, there are no changes here, we have always been using fat objects as host objects. Fat object (non LTO) is actually a host object with a few additional sections that contain target objects as data. These sections are not expected to be included into the linked image because they are marked with a special attribute that tells linker to discard them from final image.

**Review by mdtoguchi — APPROVED** [review]
LGTM

**Pull request merged**
